### PR TITLE
Firefox theme loader rework

### PR DESCRIPTION
**Pull request by ZimCodes**
This pull request includes a rework to how Doki themes are loaded. In addition, a new feature has been added called 'mixed'.

## Description

There's quite a bit of changes made to this update, so I'll state the iconic changes.

### Theme Loading Rework
As stated in issue #14,
> I finished reworking the method for loading the background images onto a new tab. This rework will provide smoother/faster transitions, as this feature takes advantage of the chrome_url_override key to load a custom page rather than forcing the browser to redirect there.

The downside is it takes a bit more to startup Firefox. Otherwise, its smooth sailing after that.

### Popup Menu
The popup menu is now smaller. A small change, but a bigger difference.

### Forgotten Images
Adds _Monika's_ & _Natsuki's_ alternate background images, and updating the _Doki Theme Definitions_ background script accordingly for each.

### New! Mixed Feature
A new feature has been added to the mix. When enabled, this feature displays a random doki theme everytime a new tab is created. Each tab has its own theme attach to them and are completely independent from each other.

## Motivation and Context

As I was looking through MDN, I found two manifest keys which will automatically load a Doki theme every new tab without the need of creating a custom method of doing so. Since it is a built-in feature, the transitions are a lot better than they were before.

After the introduction of the random feature, an idea for a different type of random theme loading appeared. A way to have different Doki themes present in a single window that is.
## How Has This Been Tested?

This has been tested using the [web-ext npm](https://github.com/mozilla/web-ext) command-line tool, which provides live updating for debugging.
## Screenshots (if appropriate):
Screenshot (53)
Screenshot (54)
Screenshot (55)
Screenshot (56)
Screenshot (57)
Screenshot (58)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-Functional Change (non-user facing changes)

#### Checklist:

- [ ] I updated the version (base package.json and re-built the master extension).
- [ ] I updated the changelog with the new functionality.

### CODE REVIEW

**Comment by Unthrottled** [comment]
Wow! Thanks again for getting this set up!

Hopefully I'll have some time to take a peek at the changes early next week or when I am adding the new themes to this repository. I'd like to keep my focus on getting the new themes out, I think they are going to turn out nice! 

Thanks for your work and understanding!

File: firefoxThemes/waifus/Aqua/theme.json
Comment (Unthrottled) [inline firefoxThemes/waifus/Aqua/theme.json:2]: Sorry about that! I forgot to take it out of the template.

File: firefoxThemes/waifus/waifu.js
Comment (Unthrottled) [inline firefoxThemes/waifus/waifu.js:23]: Ah! Thank you for putting it back into the shorthand! :100:

File: firefoxThemes/resources.js
Comment (Unthrottled) [inline firefoxThemes/resources.js:73]: Just curious, are you following some example or documentation to add this functionality?

If so, could you please share?

**Review by Unthrottled — COMMENTED** [review]
For starters, thanks again for doing this work! You know have my full attention at the moment, so I should be pretty responsive now.

There are some things that we'll need to work together on, such as the inclusion of both the primary and secondary backgrounds of each of the themes. 

Before you can do any more work, I'll need to update the base to work with the [v7.0.0 master theme release](https://github.com/doki-theme/doki-master-theme/releases/tag/v7.0.0).

I'll let you know when I have that done.

**Comment by Unthrottled** [comment]
Another thing that I noticed was that all the `doki-theme-logo@XX.png` images where replaced with the previous lower quality images (probably a merge issue). 

image


If you could please revert those changes, that'd be super!

- [ ] Logo changes reverted.

**Comment by Unthrottled** [comment]
Aight, I got the new themes in. 

I look forward to making the switch to firefox. I find it really annoying that chrome puts the doodle and search bar in the new tab. I just want to see my waifu. I think a good feature to add (eventually) would be an option to not show anything but the background.

Anyways, if you could please update to the [v7](https://github.com/doki-theme/doki-theme-chrome/tree/v7) branch, that would be awesome!

File: firefoxThemes/DokiThemeDefinitions.js
Comment (ZimCodes) [reply firefoxThemes/DokiThemeDefinitions.js:922]: 1. Mkay I'll check out the contributing file. If I don't understand something I'll let you know.

2. I see. I added these initially because I liked the secondary background of Natsuki. But if there will be more secondary backgrounds coming in... That would be cool. If secondary backgrounds are supported, they can be placed into the popup menu under the category _Alternative_ or _Secondary_ or etc.

File: firefoxThemes/resources.js
Comment (ZimCodes) [reply firefoxThemes/resources.js:73]: No. All of this comes straight from me noggin. I could explain my process for this feature if you wish.

### Inspiration
After the rework, I began testing the extension to see if everything was A-okay. What caught my interest was the random option, which randomly selects a theme for you. From here I thought, wouldn't it be cool to add to loads a random theme every time you create a new tab? Also I thought maybe someone might get tired of seeing the same waifu/theme every new tab.

### The Work
So to implement this new feature, I need to 
- listen for the current tab,
-  record the current theme for each tab, 
- listen to changes in the tab
- etc. 

With this planning cycle done, I navigate to MDN's [JavaScript API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API) section to see what I can do with the tabs. If the tabs API does not provide the tools I need to accomplish the task, there is always a workaround that can be made. An example would be the previous version of this extension. I developed my own method for loading these themes... Only to find out later there was a [simpler method](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/chrome_url_overrides).

Using the information I obtained from the docs, I formulate a plan to implement it. Sleeping helps me think. 

And that's pretty much it.

Comment (ZimCodes) [reply firefoxThemes/resources.js:73]: I'm guessing you think that way because of the excessive comments. I prefer to document what I can so when I come back I know what everything does.

File: firefoxThemes/DokiThemeDefinitions.js
Comment (Unthrottled) [reply firefoxThemes/DokiThemeDefinitions.js]: >  If secondary backgrounds are supported, they can be placed into the popup menu under the category Alternative or Secondary or etc.

I don't know if adding more things to the dropdown selection menu is best usability option. The list is starting to get pretty long. Take for instance the hyper plugin.

image

I think for the short term it could be added to the popup. Long term, if we could just give the user a list of just the distinct waifu, and allow them to set light/dark and which background they want (if supported)

File: firefoxThemes/resources.js
Comment (Unthrottled) [reply firefoxThemes/resources.js:73]: > I'm guessing you think that way because of the excessive comments

Probably, I was just curious about where you are finding out where these browser functionalities are hiding like: being able to replace the new tab action with the extensions new tab (and all of the boiler plate that came with that)

You've been adding a bunch of code, so I guess I'm trying to see if there is a place for boilerplate that you found. (Not saying boilerplate is bad, it's just stuff I don't want to write) 

For instance I just used [this guide](https://developer.chrome.com/docs/extensions/mv2/getstarted/) and [this boilerplate](https://github.com/chibat/chrome-extension-typescript-starter) to bootstrap the chrome extension.

Comment (Unthrottled) [reply firefoxThemes/resources.js:73]: Oh yeah, I appreciate you documenting your thought process. However, you don't need to write me a book 😄 You could distill it to something like this:

> I looked through the [browser JavaScript api documentation](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API), and didn't find anything, so I made my own 
> I'll do it myself
> However, I later found out the [manifest spec](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/chrome_url_overrides) supports what I wanted to do better.

File: firefoxThemes/resources.js
Comment (Unthrottled) [inline firefoxThemes/resources.js:86]: Was there a reason why each existing tab didn't get a different theme? (Did it have something to do with the background not changing?)

**Review by Unthrottled — COMMENTED** [review]
I really dig the `Mixed` mode. It's nice to be able to see all the backgrounds I've built every time I open a new tab. I put so much love and care into each of them, that it's nice to be able to see my hard work (art is hard :art:).

File: firefoxThemes/DokiThemeDefinitions.js
Comment (ZimCodes) [reply firefoxThemes/DokiThemeDefinitions.js]: Well i have some solutions to this, maybe. 

### Popup Choice
One is to pretty much do as you stated here,
> Long term, if we could just give the user a list of just the distinct waifu, and allow them to set light/dark and which background they want (if supported)

But instead keep the dropdown menu idea. 

mockup-1
The user can select the type of background to show from the left section of the popup menu. From the chosen category, the dropdown menu will be loaded with the themes in that category. So, if I choose dark, then the dropdown menu will be loaded with only dark themes.

Or the logo can be removed & the options takes its place like so:
popup-2mockup

### Sidebar Choice
The other option is a custom sidebar.
sidebar-mock
The user will first click on the doki theme icon on the top-right of the browser _as usual_, but instead of a popup menu, a sidebar will appear.

Now here's the interesting part. As you can see, the sidebar is pretty big (_disclaimer: this is the sidebar from Amazon_, which is only being used under fair use). It can pretty much host all of the Doki Themes currently available. As more themes are created, the custom sidebar can easily be adjusted to something like this.
sidebar-mock-2

Additionally, this sidebar can appear on any [valid match](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns#scheme) domain including `about:newtab`. To clarify, I can visit any site and if I want to change the theme the sidebar will always be there when needed.

File: firefoxThemes/waifus/waifu.js
Comment (ZimCodes) [reply firefoxThemes/waifus/waifu.js:33]: You could use an `async` function without `await`. `await` is used when you want something returned, but don't know how long that would be OR you want that data to be returned at some point in life. `async` is used when you want an action(_a.k.a function_) to occur, but don't care when it happens.

File: firefoxThemes/resources.js
Comment (ZimCodes) [reply firefoxThemes/resources.js:86]: Firefox does not have a built in function where each tab can have its own separate theme. There can only be one browser theme in use at a time.

Comment (ZimCodes) [reply firefoxThemes/resources.js:86]: So a workaround was made

File: firefoxThemes/waifus/waifu.js
Comment (Unthrottled) [reply firefoxThemes/waifus/waifu.js:33]: I won't labor on this point too much but, to just to make sure that we are on the same page. 

```js
browser.storage.local.get(["waifuThemes","currentThemeId"])
        .then((storage)=>{
            if(Object.keys(storage.waifuThemes.themes).includes(storage.currentThemeId)){
                addWaifu(storage);
            }
        });
```

Unless `get` returns something that is _synchronous_ and promise-like then the function passed `then` will most likely execute sometime after the `applyTheme` function. As long as `get` returns an asynchronous promise adding the `async` modifier was bit unnecessary. 

So is the  `get` method of the localstorage  synchronous or asynchronous?

File: firefoxThemes/resources.js
Comment (Unthrottled) [reply firefoxThemes/resources.js:86]: What I was trying to ask was why not assign pre-existing tabs a new random theme, so that way when the user navigates to it changes the theme. Right now when I set mixed mode it just applies the same random theme.

Peek 2020-12-30 16-15

File: firefoxThemes/DokiThemeDefinitions.js
Comment (Unthrottled) [reply firefoxThemes/DokiThemeDefinitions.js]: So I like the popup choice. 

Where it would have:

- A drop down of all the distinct waifu
- [A switch](https://material-ui.com/components/switches/#switch) to toggle light or dark (it would be disabled for a waifu that doesn't have both)
- Another switch to toggle the background between primary and secondary (it would be disabled for waifu that only have a single primary background)

However, that can be added some time after this request.

**Review by Unthrottled — APPROVED** [review]
Everything here looks fairly solid. There are some tiny things I want to work on. I'll keep you in the loop

@ZimCodes Before I merge, is there anything you are working on (should I wait?)

File: firefoxThemes/waifus/waifu.js
Comment (ZimCodes) [reply firefoxThemes/waifus/waifu.js]: You are correct.

File: firefoxThemes/resources.js
Comment (ZimCodes) [reply firefoxThemes/resources.js:86]: I see. Did not know this occurs. Thanks for pointing it out.

**Comment by ZimCodes** [comment]
Yes I need more time before the merge to fix up the mix mode issue you pointed out.

**Comment by ZimCodes** [comment]
@Unthrottled It's ready now.

File: firefoxThemes/resources.js
Comment (Unthrottled) [inline firefoxThemes/resources.js:98]: Was there a technical limitation that requires the extension to close all other tabs?

Comment (ZimCodes) [reply firefoxThemes/resources.js:98]: Yes. Since all themes load the same HTML document, only the last background image added to the local storage can be loaded. Even when multiple messages are sent from the background script to each tab's HTML page, new changes will not be applied. Even with a reload.

So two solutions:
1. Close all _new tabs_ and start from beginning
2. Create one HTML document for each background image. Each background image will have its own dedicated HTML which will load it.

Comment (Unthrottled) [reply firefoxThemes/resources.js:98]: Got it! I don't think it's that big of an issue. It just closes the empty tabs. So I'm not too concerned

File: firefoxThemes/DokiThemeDefinitions.js
Comment (Unthrottled) [reply firefoxThemes/DokiThemeDefinitions.js]: > Another switch to toggle the background between primary and secondary (it would be disabled for waifu that only have a single primary background

:point_up: I went with this for the time being, I think that it might be nice to a light dark toggle as well (I'll make an issue for that).

**Pull request merged**
